Documentation/proposal-runconfig_inside_osv.md

```diff
@@ -1,55 +1,286 @@
+*CONCERN #4*: OK, we handle runtime-specific placeholders. What about application-specific ones? E.g.
+if our NodeJS application needs to have `mongo_database_url` to be set, which is not part of node
+runtime parameters?<br>
+SOLUTION: We should prohibit application-specific arguments for run command. Instead, application
+should use environment variables for application-specific settings. To support this, we introduce a
+new attribute, `env_vars` (a list of required environment variable names) for each named configuration. E.g.
+```yaml
+# meta/run.yaml
 
+runtime: node
+
+main: app.js
+my_ip: $IP
+output_dir: $THE_OUTPUT_DIR
+env_vars:  <------------------------ the new attribute
+ - MONGO_DB_URL
+```
+Capstan would then prepend `--env key=value` option during run command creation where the value
+would be a placeholder. Run command for the example above would then become:
+```
+--env MONGO_DB_URL=___MONGO_DB_URL___ node app.js ...
+|----------- the new part -----------|
+```
+Agent would remain unchanged and would be run exactly same as before:
+```
+agent run <name-of-config> [<key-value>, <key-value>, ...]
+                            ----both: params and env ----
+
+Example:
+agent run pizza IP=192.168.0.55 THE_OUTPUT_DIR=/etc/out MONGO_DB_URL=my.mongo.local
+                |----------- params ------------------| |-------- env ------------|
+```
 
 ### Running image
-With changes above, Capstan shouldn't be required to run image. OpenStack could run it with arbitrary hypervisor as long as it provides cloud-init.
+When user decides to **RUN** an OSv image, she will have to set a proper boot command. This will be
+done using `bootcmd` cloud-init [property](https://cloudinit.readthedocs.io/en/latest/topics/examples.html#run-commands-on-first-boot). The boot command will be as simple as:
+```yaml
+# cloud-init
+bootcmd: agent run <name-of-config> [<key-value>, <key-value>, ...]
+```
+Although the boot command is simple enough to be written manually by user (and it requires no other
+arguments but the placeholder values) it will probably be written by Heat or Murano.
 
-The `capstan run` command could therefore be changed. Parameter --runconfig would result in init-file be created and passed to the OSv while run command of the composed image should remain untouched (i.e. Agent will be run).
+SIDE NOTE: We cannot rely on boot command to be set on COMPOSE time since it can easily be overwritten by
+mistake e.g. when sysadmin tests it prior uploading. This is why we can not implement an OSv Agent
+that would be run 'by default' and wait on configuration name to be passed later. Therefore this
+proposal introduces an OSv Agent that is fully defined by boot command (all parameters are there).
+
+#### Image independency
+With changes above, Capstan shouldn't be required to run image. OpenStack could run it with
+arbitrary hypervisor as long as it provides cloud-init support.
```

Comment:
Currently it reads that the hypervisor needs to provide cloud-init support. Just pointing out that hypervisor is not the one doing this.